src/bls.cpp

```diff
@@ -572,22 +591,23 @@ bool BLS::VerifyNative(
 }
 
 void BLS::HashPubKeys(relic::bn_t* output, size_t numOutputs,
-                      std::vector<BLSPublicKey> const &pubKeys) {
-    uint8_t *pkBuffer = new uint8_t[BLSPublicKey::PUBLIC_KEY_SIZE
-                                    * (pubKeys.size())];
+                      std::vector<std::vector<uint8_t>> const &serPubKeys,
+                      std::vector<size_t> const& sorted) {
     relic::bn_t order;
 
-    for (size_t i = 0; i < pubKeys.size(); i++) {
-        pubKeys[i].Serialize(pkBuffer +
-                             i * BLSPublicKey::PUBLIC_KEY_SIZE);
-    }
-
     bn_new(order);
     g2_get_ord(order);
 
+    std::vector<uint8_t> pkBuffer;
```

Comment:
Why use a vector here instead of the array?

Reply:
array seems cleaner and more consistent with the rest of the code

Reply:
And other places where arrays are replaced by vectors

Reply:
With arrays you mean dynamically allocated arrays as used before or std::array?

I prefer the use of std::vector<uint8_t> over new[] whenever possible. I actually prefer to not use new/new[] at all, simply because it is one of the main sources for memory related bugs. The memory leak that I fixed in fdf15f98ea34a8a77cc88e7b9455d4d999fa5155 was such an example. With std::vector you've generally on the secure side of things and also get a lot of flexibility (e.g. everything from std can be used, or the < which makes sorting simpler). You also have better control about ownership of buffers/arrays, as you can always pass around ownership with std::move/std::swap.

About consistency: I noticed this as well and actually started a refactoring of all places where new[] was used...but stopped in the middle as it got too much. Would you consider using std::vector in the future (for all new things) and gradually move new[] to std::vector step by step?

Reply:
With modern C++ (starting with C++11), the overhead is neglectable. Only when iterating byte-by-byte, you'd be able to measure a performance impact. But usually, you'd use stuff from std:: to perform operations, which all have specialized implementations to work fast on primitive data types.

It's however very important to pass vectors by reference, as otherwise you end up copying them over and over. Returning vectors by-value depends on the place it is used. If the vector is from the stack, the C++ compiler will will use return value optimization (avoiding a copy). If you return a member variable, it will be copied.

If you put vectors in maps or other vectors, always use m.emplace(k, vec) and v2.emplace_back(vec). If the vector is not required after adding it to another container, use m.emplace(std::move(vec)) and v2.emplace_back(std::move(vec)), which will basically swap (very fast) the contents of the inserted vector into the new position in the target container. Generally, always prefer emplace over push/insert (it avoids unnecessary copies).

If using vectors in vectors, make sure to v.reserve(expectedSize) in advance, otherwise it might end up copying stuff unnecessarily.

Reply:
@codablock do you have some backup for the negligible overhead of std::vector with C++11? 
Another alternative is to use std::unique_ptr. This avoids the memory bugs and ensures that the size is set and fixed for all time. Thus making the code easier to understand than using std::vector (across the board) where the size could be set at any place or time. With std::vector it can be harder to know whether additional elements have been added or not later.